runtime/src/bank.rs

```diff
@@ -1446,20 +1446,6 @@ impl Bank {
         self.rc
             .accounts
             .verify_hash_internal_state(self.slot(), &self.ancestors)
-            && !self.has_accounts_with_zero_lamports()
```

Comment:
(note to myself: lifting this restriction is needed because we're purging 0-lamport accounts conservatively, however is this really OK so as not to cause shielding attacks? Maybe I'm still not sure about it completely or missing something.)

Reply:
Yea, the requirement is not necessarily that there are no 0-lamport accounts, just that they do not alter the accounts view such that it would give incorrect results when the accounts db has operations performed on it. Removing all 0-lamport accounts was just an easy way to verify that there are no incorrect shielding accounts, but it wasn't correct.

We do some checks when we do verify_hash_internal_state that looks at all account paths from the root, I think we have to convince ourselves that this is enough or that we maybe need to make extra checks.

I have an idea to add another path which is like an aggressive purge_zero_lamport_accounts path which removes them from the index and then rewrites the AppendVec stores to not contain those accounts at all so that the accounts store is consistent, but I think that can be in a follow-up PR if required.

Reply:
Oh I see! Then, purging zero lamports account isn't set in stone as a hard requirement. That makes sense!